### DISCUSSION TITLE
Why is `upload_port` required in `platformio.ini`?

### DISCUSSION THREAD

**Discussion by bilogic** [arendst/Tasmota#12017] [Q&A] [answered]
I'm using VSCode with PlatformIO. When I try to comment out all the `upload_port = COM5` in `platformio.ini` and `platform_override.ini`, I get an error
```
Error: Invalid 'E:\Tasmota\platformio.ini' (project configuration file): 'No option 'upload_port' in section: 'common''
```
However, https://docs.platformio.org/en/latest/projectconf/section_env_upload.html#upload-port says that this is optional.
And google brought me to https://github.com/platformio/platformio-atom-ide/issues/1971 in which it was concluded the issue lies in `platformio.ini`.

I'm asking because in my other projects, `upload_port` does not need to be specified, PlatformIO was able to detect the correct port, especially if there is only 1 ESP8266 connected.

Any idea? Thank you.

**Comment by Jason2866** [comment, marked as answer]
Added to be sure not flashing the wrong device with autoselection of Platformio.
For testing the devs do not use just one connected esp8266 Esp32 device to the PC.
Since Tasmota uses a fine tuned Platformio setup, you can not easily remove a entry.

**Reply by bilogic** [reply]
Ok, sure. But able to point to me how it can be made optional?
Just curious as I spent almost an hour figuring out why it was not optional.

Thanks.

**Reply by Jason2866** [reply]
In short, not possible. The `upload_port` is NOT optional anymore here. 
When defining some settings or overriding, `upload_port` has to be defined too.
If you find a solution which does not break the current setup let me know.
Out of interest i spend some time in this. Never found a solution which was working and didnt break other things.

**Reply by bilogic** [reply]
Ok, I read your answer again and got it this time. I think the issue is that PlatformIO doesn't allow an `unset`, but strangely, I removed all instances of `upload_port` from all `ini` files, yet the error remained.

I will update if I find out why. Thanks.

~Hang on, do you know which part of the `ini` made `upload_port` non optional?~
~I was deciding whether to ask here or in PlatformIO as it seems like a bug to me.~
